### PR TITLE
fix(bridge): preserve no-workspace initialization

### LINKED ISSUE #742: bridge: no-workspace initialize invents a downstream CWD workspace
## Summary

When an upstream client initializes with both `rootUri: null` and `workspaceFolders: null`, Kakehashi invents its process current directory as the downstream `rootUri` and also synthesizes a one-element `workspaceFolders` array.

## Evidence

`initialize_impl` derives `primary_uri` only from the upstream workspace fields, but then builds `root_uri_for_bridge` with `std::env::current_dir()` as a fallback. The synthesized bridge root is subsequently reused to synthesize `workspace_folders_for_bridge`. Thus a downstream server observes a workspace the upstream client never supplied.

The separate internal `root_path` fallback can remain for Kakehashi config discovery; it should not change the forwarded LSP initialization contract.

## Impact

No-workspace clients (single-file sessions, embedded clients, or clients intentionally omitting a workspace) can cause downstream servers to index the directory from which Kakehashi happened to start, load unrelated project configuration, and answer requests under the wrong workspace context. This may be large or sensitive and is nondeterministic across launch environments.

## Expected behavior

Preserve the upstream no-workspace state for downstream initialization:

- downstream `rootUri` remains `null`;
- downstream `workspaceFolders` remains `null`;
- Kakehashi may still use CWD internally as its own config fallback, without forwarding it as protocol state.

## Suggested regression test

Exercise bridge initialization context derivation with both workspace fields absent and assert that both downstream values remain `None`, while separately retaining the internal root-path fallback.

### PR DESCRIPTION
**Pull request by atusy**
## Summary

- preserve an upstream no-workspace initialize state for bridged servers
- keep the process-CWD fallback internal to Kakehashi config discovery
- feed the forwarded handshake and config discovery from one client-root ladder, so a `rootPath`-only client anchors both at the workspace it named
- add regression tests for `rootUri`/`workspaceFolders` null, the legacy `rootPath` rung, and the surviving folder synthesis

Closes #742.

## Behavior

`initialize` params forwarded to bridged servers are now derived only from what the upstream client sent: `workspaceFolders[0]` → `rootUri` → deprecated `rootPath` → nothing. A client that opened no workspace forwards `rootUri: null` and no folders instead of Kakehashi's launch directory.

Kakehashi's own root keeps its CWD fallback — it anchors relative config paths and never reaches a downstream server — but it now reads the same ladder, including the `rootPath` rung the bridge honors. Previously the two agreed only by both falling back to the CWD, so a `rootPath`-only client rooted its language servers correctly while its own `kakehashi.toml` was discovered from the launch directory. The startup log names that rung instead of reporting a CWD fallback.

A relative `rootPath` still falls through to the CWD: it can only resolve against the launch directory, which is the dependence this handshake avoids.

## Scope notes

- An empty `workspaceFolders: []` is forwarded verbatim. It is the client saying it has no folders, as deliberately as a missing list; synthesizing one from `rootUri` would invent the workspace this PR removes.
- `ls_types::InitializeParams::workspace_folders` is `skip_serializing_if = "Option::is_none"`, so no folders means the key is omitted rather than sent as `null`. Both spellings are spec-legal and `rootUri: null` is emitted explicitly; expressing the distinction would need a serialization wrapper, which is out of scope here.
- Downstream servers are still spawned without an explicit `current_dir`, so one that self-falls-back to its process CWD on a null root can still land in the launch directory. That is the remaining half of #742 and needs its own change.

## Validation

- `cargo test --lib` — 3215 passed, 0 failed
- `cargo clippy --lib --tests` — no new warnings (9 pre-existing on `main`, none in the changed files)
- `cargo fmt --check`
- full local integration + e2e tier (`scripts/test_e2e_parallel.sh`) — 58 binaries, 1745 tests, 0 failures; CI does not run this tier

Downstream-side null forwarding is additionally pinned by the pre-existing `initialize_request_has_null_root_uri_when_not_provided` in `src/lsp/bridge/protocol/lifecycle.rs`.

### CODE REVIEW

File: src/lsp/lsp_impl/lifecycle.rs [hunk 1]
```diff
@@ -1597,6 +1606,18 @@ async fn upstream_forwarding_loop_with_cancel(
...
+    fn bridge_workspace_context_preserves_no_workspace_initialize() {
+        let params: InitializeParams = serde_json::from_value(serde_json::json!({
+            "capabilities": {},
+            "rootUri": null,
+            "workspaceFolders": null
+        }))
+        .expect("valid initialize params");
+
+        assert_eq!(bridge_root_uri(&params), None);
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/lifecycle.rs:2370]: Fixed in 12a7e4934. The production workspace-folder derivation is now a pure helper and the no-workspace regression asserts both downstream root URI and workspace folders remain absent.

File: src/lsp/lsp_impl/lifecycle.rs [hunk 2]
```diff
@@ -1597,6 +1606,18 @@ async fn upstream_forwarding_loop_with_cancel(
...
+    fn bridge_root_uri_preserves_no_workspace_initialize() {
+        let params: InitializeParams = serde_json::from_value(serde_json::json!({
+            "capabilities": {},
+            "rootUri": null,
+            "workspaceFolders": null
+        }))
+        .expect("valid initialize params");
+
+        assert_eq!(bridge_root_uri(&params), None);
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/lifecycle.rs]: Fixed in 12a7e4934 exactly as suggested: workspace-folder derivation is a helper used by production and the regression asserts it returns None.

File: src/lsp/lsp_impl/lifecycle.rs [hunk 3]
```diff
@@ -82,43 +128,14 @@ impl Kakehashi {
...
-                    .unwrap_or_else(|| "workspace".to_string());
-                let folder_uri = tower_lsp_server::ls_types::Uri::from_str(uri).ok()?;
-                Some(vec![tower_lsp_server::ls_types::WorkspaceFolder {
-                    uri: folder_uri,
-                    name,
-                }])
-            })
-        });
+        let workspace_folders_for_bridge =
+            bridge_workspace_folders(&params, root_uri_for_bridge.as_deref());
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/lifecycle.rs]: Fixed by deriving workspace folders before moving the root URI into the pool.

File: src/lsp/lsp_impl/lifecycle.rs [hunk 4]
```diff
@@ -54,6 +54,52 @@ fn lsp_legend_modifiers() -> Vec<SemanticTokenModifier> {
...
+fn bridge_root_uri(params: &InitializeParams) -> Option<String> {
+    #[allow(deprecated)]
+    let primary_uri = params
+        .workspace_folders
+        .as_ref()
+        .and_then(|folders| folders.first())
+        .map(|folder| &folder.uri)
+        .or(params.root_uri.as_ref());
+    primary_uri.map(|uri| uri.as_str().to_string())
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/lifecycle.rs]: Fixed: rootPath is now the final legacy fallback and has focused regression coverage.

File: src/lsp/lsp_impl/lifecycle.rs [hunk 5]
```diff
@@ -1597,6 +1614,20 @@ async fn upstream_forwarding_loop_with_cancel(
...
+        let params: InitializeParams = serde_json::from_value(serde_json::json!({
+            "capabilities": {},
+            "rootUri": null,
+            "workspaceFolders": null
+        }))
+        .expect("valid initialize params");
+
+        let root_uri = bridge_root_uri(&params);
+        assert_eq!(root_uri, None);
+        assert_eq!(bridge_workspace_folders(&params, root_uri.as_deref()), None);
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/lifecycle.rs:2369]: The extracted helpers are the initialization boundary used directly by initialize_impl, and the focused tests cover both forwarded values. Constructing a full server here would duplicate broad lifecycle setup without additional branch coverage.

**Comment by atusy** [comment]
@copilot review

**Comment by atusy** [comment]
@copilot review

File: src/lsp/lsp_impl/lifecycle.rs [hunk 6]
```diff
@@ -62,6 +62,59 @@ fn host_position_encoding(capabilities: &ClientCapabilities) -> Option<PositionE
         .map(|_| PositionEncodingKind::UTF16)
 }
 
+/// Derive only the root URI the upstream client supplied for downstream
+/// initialization. Kakehashi may use its process CWD internally for config
+/// discovery, but forwarding that fallback would turn a no-workspace session
+/// into an unrelated workspace for every bridged server.
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/lifecycle.rs]: Fixed by documenting all upstream workspace inputs, including legacy rootPath.

File: src/lsp/lsp_impl/lifecycle.rs [hunk 7]
```diff
@@ -1632,6 +1656,36 @@ mod tests {
...
+            "rootUri": null,
+            "rootPath": "/tmp/legacy-workspace",
+            "workspaceFolders": null
+        }))
+        .expect("valid initialize params");
+
+        assert_eq!(
+            bridge_root_uri(&params).as_deref(),
+            Some("file:///tmp/legacy-workspace")
+        );
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/lifecycle.rs]: Fixed by deriving both rootPath and expected URI from the platform current directory.

**Comment by atusy** [comment]
@copilot review

File: src/lsp/lsp_impl/lifecycle.rs [hunk 8]
```diff
@@ -62,6 +62,60 @@ fn host_position_encoding(capabilities: &ClientCapabilities) -> Option<PositionE
...
+        .and_then(|folders| folders.first())
+        .map(|folder| &folder.uri)
+        .or(params.root_uri.as_ref());
+    primary_uri.map(|uri| uri.as_str().to_string()).or_else(|| {
+        params
+            .root_path
+            .as_ref()
+            .and_then(|path| Url::from_file_path(path).ok())
+            .map(|uri| uri.to_string())
+    })
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/lifecycle.rs]: Valid, and fixed in 9f72e4fce.

The two roots now read one ladder (`ClientRoot`: workspaceFolders[0] → rootUri → rootPath), so a rootPath-only client anchors both the forwarded handshake and Kakehashi's own config discovery at the workspace it named. Previously they agreed only by both falling back to the CWD, which meant `/foo/kakehashi.toml` was never read while downstream servers were correctly rooted at `/foo`. The startup log gained the matching `root_path (deprecated)` arm.

A relative rootPath still falls through to the CWD fallback — it can only resolve against the launch directory, which is the dependence this PR removes — and that is pinned by `client_root_rejects_a_relative_legacy_root_path`.

**Comment by atusy** [comment]
@coderabbitai review
@greptileai
@codex review
/agentic_review

Rebased onto main (#738 had landed) and revised. Since the last round: the forwarded root and Kakehashi's own config root now read one `ClientRoot` ladder, which closes the rootPath/root mismatch; the legacy-rootPath test no longer uses the CWD as its fixture (it was exactly the value the removed fallback produced, so it pinned nothing); the surviving folder-synthesis branch gained coverage.

Deliberately unchanged, with reasoning in the PR body: `workspaceFolders: []` is forwarded verbatim, the absent-vs-null spelling of empty folders, and the `Url` vs strict-`Uri` percent-encoding difference for paths containing `[ ] | ^` (swapping encoders would move the forwarded root for ordinary paths and Windows drive letters — follow-up, not this PR).

Full local integration + e2e tier: 1744 tests, 58 binaries, 0 failures.

**Comment by atusy** [comment]
@greptileai

File: src/lsp/lsp_impl/lifecycle.rs [hunk 9]
```diff
@@ -1865,6 +1964,205 @@ mod tests {
...
+            "capabilities": {},
+            "rootUri": null,
+            "workspaceFolders": null
+        }))
+        .expect("valid initialize params");
+
+        let root_uri = bridge_root_uri(&params);
+        assert_eq!(root_uri, None);
+        assert_eq!(bridge_workspace_folders(&params, root_uri.as_deref()), None);
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/lifecycle.rs:2370]: Taken, and fixed in 3cb8de481 + af589fd67.

You are right that only half the contract was pinned. The internal fallback was inline in `initialize_impl`, so asserting it meant driving the whole initialize seam (which reads `$HOME`/XDG state). Extracting `config_root_path` made it testable directly, and three tests now pin it: the no-workspace case asserts in one place that the handshake forwards `None` while config discovery still resolves the CWD, plus the two log-source rungs — a client-supplied `rootPath` must not be reported as a CWD fallback, and an unanchorable one must be.

3203 lib tests green.
